ompi/mca/common/ompio/common_ompio_request.c

```diff
@@ -142,16 +150,66 @@ int mca_common_ompio_progress ( void )
-                completed++;
-                ompi_request_complete (&req->req_ompi, true);
+        if (NULL != req->req_progress_fn) {
+            if (req->req_progress_fn(req)) {
+                /**
+                 * To support pipelined read/write operations, a user level request
+                 * can contain multiple internal requests. These sub-requests
+                 * contain a pointer to the parent request.
+                 */
+                mca_ompio_request_t *parent = req->req_parent;
+                if (NULL != parent) {
+                    /* This is a subrequest */
+                    if (OMPI_SUCCESS != req->req_ompi.req_status.MPI_ERROR) {
+                        parent->req_ompi.req_status.MPI_ERROR = req->req_ompi.req_status.MPI_ERROR;
+                        ompi_request_complete (&parent->req_ompi, true);
+                        continue;
+                    }
+                    parent->req_subreqs_completed++;
+                    parent->req_ompi.req_status._ucount += req->req_ompi.req_status._ucount;
+                    req->req_post_followup = true;
+                } else {
+                    /* This is a request without subrequests */
+                    completed++;
+                    ompi_request_complete (&req->req_ompi, true);
+                }
                 /* The fbtl progress function is expected to set the
                  * status elements
                  */
             }
+        } else {
+            /* This is a request without a lower level progress function, .e.g
+             * a parent request
+             */
+            if (req->req_num_subreqs == req->req_subreqs_completed) {
+                completed++;
+                ompi_request_complete (&req->req_ompi, true);
+            }
         }
+    }
 
+    /**
+     * Splitting the ompio progress loop is necessary to avoid that a pending operation
+     * consisting of multiple subrequests is executed in a single invokation of the progress
+     * function.
+     *
+     * Otherwise it can happen that the next subrequest is posted, which ends up at the tail
+     * of the ompio_pending_requests_list, and would be processed in the same loop execution;
+     * which then posts the next subrequest, which is also processed potentially right away
+     * etc. This would make the ompio_progress function block for a long time, and prevent
+     * overlapping operations.
+     *
+     * Splitting the loop into two parts, one checking for completion and one posting
+     * the next subrequest if necessary avoids the problem.
+     */
+    OPAL_LIST_FOREACH(litem, &mca_common_ompio_pending_requests, opal_list_item_t) {
+        req = GET_OMPIO_REQ_FROM_ITEM(litem);
+        if (true == req->req_post_followup) {
+            mca_ompio_request_t *parent = req->req_parent;
+            parent->req_post_next_subreq(parent, parent->req_subreqs_completed);
+            ompi_request_complete (&req->req_ompi, true);
```

Comment:
Will there be someone waiting on this request? Otherwise passing `true` here is not necessary (IIRC)